[Package]: spirv-tools

### Why is it worth to add this package?

For glslang (from Termux repo,)
and to allow Vulkan compilation (terminal emulator's main purpose = compilation.)
`~/Vulkan-Hpp $ cmake --build .
[0/1] Re-running CMake...
--  Found clang-format version <16.0.6>.
--  Using .clang-format version 15.
CMake Deprecation Warning at glm/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


CMake Deprecation Warning at glm/CMakeLists.txt:2 (cmake_policy):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Including X11 support
CMake Deprecation Warning at glslang/CMakeLists.txt:36 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


CMake Warning (dev) at glslang/CMakeLists.txt:251 (find_package):
  Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
  are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

Call Stack (most recent call first):
  glslang/CMakeLists.txt:315 (find_host_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Google Mock was not found - tests based on that will not build
-- spirv-tools not linked - illegal SPIRV may be generated for HLSL`
https://duckduckgo.com/?q=%22spirv-tools++++not+linked%22
result 1:
https://bugs.archlinux.org/task/69974

### Home page URL

_No response_

### Source code URL

https://github.com/KhronosGroup/SPIRV-Tools

### Packaging policy acknowledgement

- [X] I certify that I have read [Termux Packaging Policy](https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#packaging-policy) and understand that my request will be denied in case of violation.

### Additional information

_No response_

Worse after update, now breaks compilation:
`~/Vulkan-Hpp $ cmake -DVULKAN_HPP_SAMPLES_BUILD=ON -DVULKAN_HPP_TESTS_BUILD=ON -G Ninja

CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


-- The CXX compiler identification is Clang 16.0.6
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /data/data/com.termux/files/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
--  Found clang-format version <16.0.6>.
--  Using .clang-format version 15.
-- GLM: Version 0.9.9.9
-- The C compiler identification is Clang 16.0.6
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /data/data/com.termux/files/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found Doxygen: /data/data/com.termux/files/usr/bin/doxygen (found version "1.9.8") found components: doxygen
-- Including X11 support
-- Found X11: /data/data/com.termux/files/usr/include
-- Looking for XOpenDisplay in /data/data/com.termux/files/usr/lib/libX11.so;/data/data/com.termux/files/usr/lib/libXext.so
-- Looking for XOpenDisplay in /data/data/com.termux/files/usr/lib/libX11.so;/data/data/com.termux/files/usr/lib/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Looking for IceConnectionNumber in ICE
-- Looking for IceConnectionNumber in ICE - found
-- Found Python3: /data/data/com.termux/files/usr/bin/python3.11 (found version "3.11.6") found components: Interpreter
-- Google Mock was not found - tests based on that will not build
CMake Error at glslang/CMakeLists.txt:280 (message):
  ENABLE_OPT set but SPIR-V tools not found.  Please run
  update_glslang_sources.py, set the ALLOW_EXTERNAL_SPIRV_TOOLS option to use
  a local install of SPIRV-Tools, or set ENABLE_OPT=0.


-- optimizer enabled
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Vulkan: /data/data/com.termux/files/usr/lib/libvulkan.so (found version "1.3.266") found components: glslangValidator missing components: glslc
-- Configuring incomplete, errors occurred!
`

--- a/packages/glslang/build.sh
+++ b/packages/glslang/build.sh
@@ -3,11 +3,23 @@ TERMUX_PKG_DESCRIPTION="OpenGL and OpenGL ES shader front end and validator"
 TERMUX_PKG_LICENSE="BSD"
 TERMUX_PKG_MAINTAINER="@termux"
 TERMUX_PKG_VERSION="13.0.0"
-TERMUX_PKG_SRCURL=https://github.com/KhronosGroup/glslang/archive/${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_REVISION=1
+TERMUX_PKG_SRCURL=https://github.com/KhronosGroup/glslang/archive/refs/tags/${TERMUX_PKG_VERSION}.tar.gz
 TERMUX_PKG_SHA256=bcda732434f829aa74414ea0e06d329ec8ac28637c38a0de45e17c8fd25a4715
 TERMUX_PKG_DEPENDS="libc++"
+TERMUX_PKG_BUILD_DEPENDS="spirv-tools"
 TERMUX_PKG_NO_STATICSPLIT=true
 TERMUX_PKG_AUTO_UPDATE=true
 TERMUX_PKG_EXTRA_CONFIGURE_ARGS="
--DBUILD_SHARED_LIBS=ON
+-DALLOW_EXTERNAL_SPIRV_TOOLS=ON
 "
+
+termux_step_post_make_install() {
+	# build system only build static or shared at a time
+	TERMUX_PKG_EXTRA_CONFIGURE_ARGS+="
+	-DBUILD_SHARED_LIBS=ON
+	"
+	termux_step_configure
+	termux_step_make
+	termux_step_make_install
+}

--- a/packages/shaderc/build.sh
+++ b/packages/shaderc/build.sh
@@ -1,0 +1,19 @@
+TERMUX_PKG_HOMEPAGE=https://github.com/google/shaderc
+TERMUX_PKG_DESCRIPTION="Collection of tools, libraries, and tests for Vulkan shader compilation"
+TERMUX_PKG_LICENSE="Apache-2.0"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION="2023.6"
+TERMUX_PKG_SRCURL=https://github.com/google/shaderc/archive/refs/tags/v${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=e40fd4a87a56f6610e223122179f086d5c4f11a7e0e2aa461f0325c3a0acc6ae
+TERMUX_PKG_DEPENDS="libc++"
+TERMUX_PKG_CONFLICTS="glslang, spirv-tools"
+TERMUX_PKG_NO_STATICSPLIT=true
+TERMUX_PKG_AUTO_UPDATE=true
+TERMUX_PKG_UPDATE_TAG_TYPE="newest-tag"
+TERMUX_PKG_EXTRA_CONFIGURE_ARGS="
+-DSHADERC_SKIP_TESTS=ON
+"
+
+termux_step_post_get_source() {
+	./utils/git-sync-deps
+}

--- a/packages/spirv-headers/build.sh
+++ b/packages/spirv-headers/build.sh
@@ -1,0 +1,32 @@
+TERMUX_PKG_HOMEPAGE=https://github.com/KhronosGroup/SPIRV-Headers
+TERMUX_PKG_DESCRIPTION="SPIR-V Headers"
+TERMUX_PKG_LICENSE="Apache-2.0"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION="1.3.261.1"
+TERMUX_PKG_SRCURL=https://github.com/KhronosGroup/SPIRV-Headers/archive/refs/tags/sdk-${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=32b4c6ae6a2fa9b56c2c17233c8056da47e331f76e117729925825ea3e77a739
+TERMUX_PKG_PLATFORM_INDEPENDENT=true
+TERMUX_PKG_AUTO_UPDATE=true
+
+termux_pkg_auto_update() {
+	local e=0
+	local api_url="https://api.github.com/repos/KhronosGroup/SPIRV-Headers/git/refs/tags"
+	local api_url_r=$(curl -s "${api_url}")
+	local r1=$(echo "${api_url_r}" | jq .[].ref | sed -nE "s|.*/(sdk-.*)\"|\1|p")
+	local latest_version=$(echo "${r1}" | tail -n1)
+	[[ -z "${api_url_r}" ]] && e=1
+	[[ -z "${r1}" ]] && e=1
+	[[ -z "${latest_version}" ]] && e=1
+
+	if [[ "${e}" != 0 ]]; then
+		cat <<- EOL >&2
+		WARN: Auto update failure!
+		api_url_r=${api_url_r}
+		r1=${r1}
+		latest_version=${latest_version}
+		EOL
+		return
+	fi
+
+	termux_pkg_upgrade_version "${latest_version//sdk-}"
+}

--- a/packages/spirv-tools/build.sh
+++ b/packages/spirv-tools/build.sh
@@ -1,0 +1,39 @@
+TERMUX_PKG_HOMEPAGE=https://github.com/KhronosGroup/SPIRV-Tools
+TERMUX_PKG_DESCRIPTION="SPIR-V Tools"
+TERMUX_PKG_LICENSE="Apache-2.0"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION="1.3.261.1"
+TERMUX_PKG_SRCURL=https://github.com/KhronosGroup/SPIRV-Tools/archive/refs/tags/sdk-${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=ead95c626ad482882a141d1aa0ce47b9453871f72c42c0b28d39c82f60a52008
+TERMUX_PKG_DEPENDS="libc++"
+TERMUX_PKG_BUILD_DEPENDS="spirv-headers (=${TERMUX_PKG_VERSION})"
+TERMUX_PKG_NO_STATICSPLIT=true
+TERMUX_PKG_AUTO_UPDATE=true
+TERMUX_PKG_EXTRA_CONFIGURE_ARGS="
+-DSPIRV-Headers_SOURCE_DIR=${TERMUX_PREFIX}
+-DSPIRV_WERROR=OFF
+"
+
+termux_pkg_auto_update() {
+	# use versions from sdk-* and not vYEAR.* to match spirv-headers
+	local e=0
+	local api_url="https://api.github.com/repos/KhronosGroup/SPIRV-Tools/git/refs/tags"
+	local api_url_r=$(curl -s "${api_url}")
+	local r1=$(echo "${api_url_r}" | jq .[].ref | sed -nE "s|.*/(sdk-.*)\"|\1|p")
+	local latest_version=$(echo "${r1}" | tail -n1)
+	[[ -z "${api_url_r}" ]] && e=1
+	[[ -z "${r1}" ]] && e=1
+	[[ -z "${latest_version}" ]] && e=1
+
+	if [[ "${e}" != 0 ]]; then
+		cat <<- EOL >&2
+		WARN: Auto update failure!
+		api_url_r=${api_url_r}
+		r1=${r1}
+		latest_version=${latest_version}
+		EOL
+		return
+	fi
+
+	termux_pkg_upgrade_version "${latest_version//sdk-}"
+}

